Feature/decouple thumbnail downloads and cache

schroda opened the pull request:
Currently all thumbnails are downloaded permanently.
This not only includes thumbnails of mangas that are part of the library but also when browsing a source.
Unless you manually cleanup these files, they will never get deleted.

This PR changes thumbnails to be tmp files, which get saved in the systems tmp folder.
Only thumbnails of mangas, that are part of the library, will be saved permanently.
Once a manga gets added/removed from the library, the permanent thumbnail will get downloaded/deleted.

The logic for this reuses the existing logic of the chapter pages.
There had to be slight adaptions to the interface and classes since the functions for the thumbnails and pages have different number of arguments.
Since they are still classes which have to provide the same behaviour with the same functions (e.g. get image, download image, delete image) I wanted to use the same interface

In case the tmp files should not get saved in the systems tmp folder, this can be changed in a later/separate PR which implements a custom cache (which IMO then should also be used for the pages of chapters)

schroda committed:
Rename "DownloadedFilesProvider" to "ChaptersFilesProvider"

schroda committed:
Further abstract "DownloadedFilesProvider"

schroda committed:
Rename "getCachedImageResponse" to "getImageResponse"

schroda committed:
Extract getting cached image response into new function

schroda committed:
Move thumbnail cache to actual temp folder

schroda committed:
Rename "mangaDownloadsRoot" to "downloadRoot"

schroda committed:
Move manga downloads into "mangas" subfolder

AriaMoradi commented on server/src/main/kotlin/suwayomi/tachidesk/manga/impl/util/storage/ImageResponse.kt:
```diff
@@ -48,13 +56,9 @@ object ImageResponse {
         val filePath = "$saveDir/$fileName"
 
-        // in case the cached file is a ".tmp" file something went wrong with the previous download, and it has to be downloaded again
+// in case the cached file is a ".tmp" file something went wrong with the previous download, and it has to be downloaded again
```
fix indentation

AriaMoradi commented on server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Chapter.kt:
```diff
@@ -331,13 +331,24 @@ object Chapter {
         }
     }
 
+    private fun deleteFilesOf(chapterIds: List<Int>, mangaId: Int) {
```
choose a better name for this function

schroda replied:
renamed it to `deleteDownloadsFor`

AriaMoradi commented on server/src/main/kotlin/suwayomi/tachidesk/manga/impl/download/fileProvider/IDownloadedFilesProvider.kt:
```diff
@@ -0,0 +1,5 @@
+package suwayomi.tachidesk.manga.impl.download.fileProvider
+
+interface IDownloadedFilesProvider : IFileDownloader, IFileRetriever {
```
there is no need to include "I" in front of interface names

AriaMoradi commented on server/src/main/kotlin/suwayomi/tachidesk/manga/impl/download/fileProvider/IFileDownloader.kt:
```diff
@@ -0,0 +1,29 @@
+}
+
+@FunctionalInterface
+interface IFileDownload1<A> : IFileDownload {
```
make up better names instead of numbering

schroda replied:
tbh, no clue how to name this, I added "Args" as an suffix
I'm open for better names

AriaMoradi commented on server/src/main/kotlin/suwayomi/tachidesk/manga/impl/util/DirName.kt:
```diff
@@ -44,8 +44,12 @@ private fun getChapterDir(mangaId: Int, chapterId: Int): String {
+
 fun getChapterDownloadPath(mangaId: Int, chapterId: Int): String {
-    return applicationDirs.mangaDownloadsRoot + "/" + getChapterDir(mangaId, chapterId)
+    return applicationDirs.downloadsRoot + "/mangas/" + getChapterDir(mangaId, chapterId)
```
bad abstraction here

schroda replied:
added `mangaDownloadsRoot` to app dirs

AriaMoradi replied:
also we should use a sensible path joiner function instead of hand concatenating them

schroda committed:
Move files into sub packages

schroda committed:
Decouple thumbnail cache and download

schroda committed:
Download and delete permanent thumbnails

When adding/removing manga from/to the library make sure the permanent thumbnail files will get handled properly

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/graphql/mutations/MangaMutation.kt:
```diff
@@ -71,7 +82,7 @@ class MangaMutation {
     }
 
-    fun updateMangas(input: UpdateMangasInput): UpdateMangasPayload {
+    suspend fun updateMangas(input: UpdateMangasInput): UpdateMangasPayload {
```
Public functions in the `mutations` and `queries` packages cannot be suspend, but they can return a `CompleteableFuture` created by the `future {}` function

schroda replied:
fixed

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/graphql/mutations/MangaMutation.kt:
```diff
@@ -44,46 +45,60 @@ class MangaMutation {
     )
 
-    private fun updateMangas(ids: List<Int>, patch: UpdateMangaPatch) {
+    private suspend fun updateMangas(ids: List<Int>, patch: UpdateMangaPatch) {
```
Can you revert most of the changes here? I had it like this so that we could add more user fields with minimal changes, just do another check for inLibrary when choosing what to do with the thumbnail

schroda replied:
done

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/manga/impl/util/storage/SafePath.kt:
```diff
@@ -9,6 +11,10 @@ package suwayomi.tachidesk.manga.impl.util.storage
 
 // adopted from: https://github.com/tachiyomiorg/tachiyomi/blob/4cefbce7c34e724b409b6ba127f3c6c5c346ad8d/app/src/main/java/eu/kanade/tachiyomi/util/storage/DiskUtil.kt
 object SafePath {
+    fun joinPaths(vararg paths: String): String {
```
I'm not too sure about this function, I would prefer just doing something like this instead of using a function

```kotlin
val path = Path(System.getProperty("java.io.tmpdir")) / "Tachidesk"
```

schroda replied:
so, just leave it as it was? https://github.com/Suwayomi/Tachidesk-Server/pull/581#discussion_r1240983810

Syer10 replied:
Path.div is an operator function for Path objects

schroda replied:
going a step back, does this actually matter (regarding the sensible path joiner)? won't it be correct either way once the path is used to create a File or Path?

Syer10 replied:
Nope, it doesn't matter. I just found the joinPaths function weird, if you want to revert those changes and just use / that's fine, since / is supported by every platform

schroda replied:
I reverted it

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Manga.kt:
```diff
@@ -284,8 +286,23 @@ object Manga {
+
     private fun clearMangaThumbnailCache(mangaId: Int) {
-        val saveDir = applicationDirs.thumbnailsRoot
+        val saveDir = applicationDirs.tempThumbnailCacheRoot
```
We call this when the thumbnail url was changed, so we need to download the new cover, is this properly handled for library manga?

schroda replied:
it was not - updated

schroda committed:
Clear downloaded thumbnail
